### ISSUE TITLE
Naming of mutators

### ISSUE THREAD

**Issue by richardwerkman** [stryker-mutator/stryker-handbook#11]
Currently the list of mutators says:

- Binary Operators
- Boolean Substitutions
- Logical operators
- Unary operators
- Update operators
- Remove conditionals
- Assignment mutator
- Array declarator
- String mutator
- Block statement
- Checked mutator
- Method mutator

It is pretty inconsistant. Some have "mutator" in their name, others don't. Some say what the mutator does (Remove conditionals), others just describe what kind of operators they target. 

I think we should make this list more consistant, and use a format for each mutator name. 

Maybe all should be called mutator? Or all should describe what kind of statements/expressions they mutate. What do you guys think?

**Comment by nicojs** [comment]
> Binary operators is by far the biggest, while in my mind `a + b` and `a > b` are quite different

Agreed. What do you think of this:

* Arithmetic Operators: + - * / %
* Compare Operators:  <, >, >=, <=
* Equality Operators: ==, !==

Than it will fit nicely with "Logical Operators" 

It is worth noting, if we are going to call the Mutators by the type of node they are mutating, that the technical name can differ between the different kind of AST's (Scala, JS, TypeScript and C#). For example: a `ArithmeticOperatorNode` does not exist in the JS/TS AST.  We should all agree on the name here and implement them as such, regardless of technical names of the AST.

All in all this should be the new list in my opinion:

* ~~Binary operators~~
* *Arithmetic operator*
* *Compare operator*
* *Equality operator*
* Logical operator
* Boolean Literal
* Unary operator
* Update operator
* ~~Remove conditionals~~ *Control statement*
* Assignment ~~mutator~~ *expression*
* Array ~~declarator~~ *literal* 
* String *literal*
* Block statement
* Checked ~~mutator~~ *statement*
* Method ~~mutator~~ *expression*

**All singular, without the mutator postfix** (I think it does make sense for the class names that implement them to have the `Mutator` post fix)

@hugo-vrijswijk @simondel @richardwerkman do you agree?

**Comment by nicojs** [comment]
"Control statements" as in [Control flow](https://en.wikipedia.org/wiki/Control_flow) or [Control statement](http://www.tutorialspoint.com/ansi_c/c_control_statements.htm). 

I feel a "Conditional expression" is more the part inside the brackets of the for/while/if statement.

**Comment by hugo-vrijswijk** [comment]
Yeah we should definitely keep the names non-language specific (the node names are pretty general in Scala, compare, arithmetic, method are all the same, it's just syntax sugar). 

Compare and equality are also similar, in my opinion (and we even mutate them to each other), so I'd put those together.

> I feel a "Conditional expression" is more the part inside the parenthesis of the for/while/if statement.

[But isn't that exactly what the mutators are doing?](https://github.com/stryker-mutator/stryker-handbook/blob/master/mutator-types.md#remove-conditionals)

**Comment by nicojs** [comment]
> Compare and equality are also similar, in my opinion (and we even mutate them to each other), so I'd put those together.

So the name should be "Equality Operator" ?

> But isn't that exactly what the mutators are doing?

Wow I guess you're right. So let's call it "Conditional expression". @richardwerkman do you want to update your original issue, so we can view the current mutator proposal names?

**Comment by richardwerkman** [comment]
@nicojs @hugo-vrijswijk I have updated the list in the issue to fit your comments. I'm still doubtfull about the `Method expression`, since it is not an actual expression, more of a identifier for the method to invoke. It could become `Method call` or `Method identifier`. What is your thought?

**Comment by nicojs** [comment]
Good point. We should maybe group the methods as well. Otherwise that group might become too big. @hugo-vrijswijk any thoughts on how to group them? Since Stryker4s will probably have the most of these mutator? For C#, we might be able to group all LINQ expression mutations in a `Linq Method Call`.

**Comment by hugo-vrijswijk** [comment]
`Method Call` I think is a good descriptive name. Most of the current method mutators for Stryker4s are on collections/monads and I don't think those can be split up much more. I don't think we should make categories that are language-specific like `Linq method call` or something. It might get a little big, but anything under, say, 30 mutators still has overview, in my opinion

**Comment by legopiraat** [comment]
Is there a definitive list yet?

**Comment by nicojs** [comment]
I think so. Who makes the PR?

**Comment by richardwerkman** [comment]
Stryker.NET already implemented the list. So yeah making this list definite seems right for me!

**Comment by hugo-vrijswijk** [comment]
One final suggestion, looking at the list. `Array declarator literal` is a bit long. Maybe we could change it to just `Array Declaration`?
